man/johab2syllables.1
.TH JOHAB2SYLLABLES 1 "30 July 2023"
.SH NAME
johab2syllables \- Convert Unifont Johab 6/3/1 file to Hangul Syllables
.SH SYNOPSIS
\fBjohab2syllables \-i\fP \fIhangul-base.hex \fP
\fB\-o\fP \fIhangul-syllables.hex\fP
.SH DESCRIPTION
.B johab2syllables
generates the modern Hangul Syllables Unicode range, U+AC00..U+D7A3,
from an input Unifont .hex file encoded in Johab 6/3/1 format.  This
program demonstrates how to use functions in the unihangul-support.c
file to map a Hangul Syllables code point to the choseong, jungseong,
and jongseong Hangul Jamo that comprise each syllable.
.PP
This program is specifically designed to generate the Unicode
Hangul Syllables range only, and does \fInot\fP shift final consonants
(jongseong) if the vowel contains two long vertical strokes.
For a more general program that uses the entire Hangul Jamo
modern, ancient, and extended ancient ranges of U+1100..U+11FF,
U+A960..U+A97F, and U+D7B0..D7FF, and which \fIdoes\fP shift
jongseong right one column for vowels with two long vertical
strokes on the right, see
.BR unigen-hangul (1).
That program is  used to generate the distributed Unifont
Hangul Syllables range.
.RS
.IP \fBRange\fP 15
\fBHangul\fP
.TP
.BR 1100-1112
Modern initial consonants (choseong)
.TP
.BR 1161-1175
Modern medial vowels and diphthongs (jungseong)
.TP
.BR 11A8-11C2
Modern final consonants (jongseong).
.RE
.PP
.SH OPTIONS
.RS
.IP \fBOption\fP 15
\fBFunction\fP
.TP
\fB\-h, \-\-help\fP
Print a help message and exit.
.TP
\fB\-i\fP \fIinput_file\fP
Unifont hangul-base.hex formatted input file.
.TP
\fB\-o\fP \fIoutput_file\fP
Unifont .hex format output file.
.RE
.SH EXAMPLES
.TP 5
.RS
johab2syllables -i hangul-base.hex -o hangul-syllables.hex
.PP
johab2syllables < hangul-base.hex > hangul-syllables.hex
.RE
.SH FILES
Unifont .hex files in Johab 6/3/1 encoding.  See
.BR unifont-johab631 (5)
for a description of the input file structure.  This program
uses functions contained in the file unihangul-support.c.
.SH SEE ALSO
.BR bdfimplode (1),
.BR hex2bdf (1),
.BR hex2otf (1),
.BR hex2sfd (1),
.BR hexbraille (1),
.BR hexdraw (1),
.BR hexkinya (1),
.BR hexmerge (1),
.BR johab2ucs2 (1),
.BR unibdf2hex (1),
.BR unibmp2hex (1),
.BR unibmpbump (1),
.BR unicoverage (1),
.BR unidup (1),
.BR unifont (5),
.BR unifont-johab631 (5),
.BR unifont-viewer (1),
.BR unifont1per (1),
.BR unifontchojung (1),
.BR unifontksx (1),
.BR unifontpic (1),
.BR unigen-hangul (1),
.BR unigencircles (1),
.BR unigenwidth (1),
.BR unihex2bmp (1),
.BR unihex2png (1),
.BR unihexfill (1),
.BR unihexgen (1),
.BR unihexpose (1),
.BR unihexrotate (1),
.BR unijohab2html (1),
.BR unipagecount (1),
.BR unipng2hex (1)
.SH AUTHOR
.B johab2syllables
was written by Paul Hardy.
.SH LICENSE
.B johab2syllables
is Copyright \(co 2023 Paul Hardy.
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.
.SH BUGS
No known bugs exist.


NEWS
2024-02-24: Release 15.1.05<F2>
  * Added almost 600 CJK ideographs, including:
    - The entire CJK Unified Ideographs Extension D range of 222 glyphs
    - 335 commonly used Cantonese ideographs in Plane 2 and Plane 3.
    - Several updates to CJK Unified Ideographs Extension I.
  * All Hangul glyphs now have a consistent look with the Hangul Syllables,
    with new glyyphs for:
    - U+1100..U+11FF Hangul Jamo
    - U+3130..U+318F Hangul Compatibility Jamo
    - U+A960..U+A97F Hangul Jamo Extended-A
    - U+D7B0..U+D7FF Hangul Jamo Extended-B.
  * Several other glyph modifications; see the ChangeLog file for details.

2023-10-29: Release 15.1.04
  * Added larger Plane 2 supersets to default and Japanese font files.

2023-10-21: Release 15.1.03
  * Boris Zhang and Yzy32767 contributed CJK Unified Ideographs
    Extension I glyphs (U+2EBF0..U+2EE5D).
  * 湖 远星 contributed 14 glyph updates to CJK Unified Ideographs
    Extension B and Extension C: U+2A6D7..U+2A6DF and
    U+2B735..U+2B739.

2023-09-21: Release 15.1.02
   * 湖 远星 adjusted 46 glyphs in the Wen Quan Yi range, U+2F00..U+9FFF, and
     contributed CJK Unified Ideographs Extension G glyphs U+30000..U+3017F.
   * Added new man pages to MAN1PAGES & MAN5PAGES definitions for installation.

2023-09-12: Release 15.1.01
   * Glyphs updated for Unicode 15.1.0 release.
   * OpenType fonts are now built by default, replacing TrueType fonts.
     TrueType fonts can be built manually by invoking "make truetype"
     in the font directory.
   * New Hangul Johab 6/3/1 encoding devised by Ho-Seok Ee for building
     the Hangul Syllables block (U+AC00..U+D7A3), with new supporting
     programs johab2syllables, unigen-hangul, and unijohab2html.  This
     encoding replaces Unifont's original Hanterm-derived encoding.
     New source files hangul.h and unihangul-support.c provide support
     for rendering syllable bitmaps with all Unicode Hangul letters,
     including all ancient Hangul, through the new unigen-hangul program.
     These functions also can support user-written Hangul syllable rendering
     programs, for example on embedded systems using graphics displays.
   * Added three hexadecimal digit representation CSUR/UCSUR scripts:
     - U+EBE0..U+EBEF: Boby Lapointe's "bibi-binary" notation
     - U+EBF0..U+EBFF: Bruce Alan Martin's bit location notation
     - U+ECF0..U+ECFF: Ronald O. Whitaker's triangular notation.
   * Modified hexdraw to support bitmap manipulation of up to 64 rows
     by 128 columns for 64-by-128 graphics display screen drawing support.
     Added the unihexpose program to transpose Unifont glyphs, for use
     with graphics display controller chips that encode each byte as
     8 rows by 1 column, left to right.
   * Other minor updates; see ChangeLog for details.

2023-06-04: Release 15.0.06
   * Further adjustments to Hangul glyphs in the Basic Multilingual Plane.

2023-06-03: Release 15.0.05
   * New Hangul Syllables block (U+AC00..U+D7A3) contributed
     by Minseo Lee.  Generated with his modifications of three
     combining patterns in the johab2ucs2 Perl script.
   * Modified U+5829 and U+6FF9 Chinese glyphs.
   * Full details in ChangeLog file.

2023-05-28: Release 15.0.04
   * Adjustments to Greek and Coptic letters.
   * Adjustments to two heart glyphs.
   * Added Engsvanáli (U+E100..U+E14F) CSUR glyphs.

2023-05-21: Release 15.0.03
   * Fixed a typo in src/Makefile that could affect
     "make install".
   * Install unifont_all.hex in $(PKGDEST) directory.
   * Other minor changes; see ChangeLog file for details.

2023-05-20: Release 15.0.02
   * N.B.: This is the last planned version to build TrueType
     font files as part of the standard font build with

           make BUILDFONT=1

     In the future, only OpenType font files will be part of
     the ordinary font building using the hex2otf program.
     TrueType fonts will have to be built manually using

           cd font
           make truetype

   * Minor update to allow for separate building of TrueType
     and OpenType fonts.  This is in preparation for the future
     removal of TrueType font files from default font builds,
     in preference for OpenType font files.  That change is
     planned for the release of Unifont 15.1.01, which will
     correspond to the release of the Unicode 15.1.0 standard
     scheduled for September 2023.
   * Added Doxygen annotation to C programs and corresponding
     ".h" header files.  Annotated output is generated in a
     new "doxygen" subdirectory.  Doxygen output consists of
     a browsable set of HTML pages and a hyperlinked PDF document.
   * Updates to make files to fix issues with redundant building
     under some circumstances.
   * Added 196 Plane 2 and Plane 3 Chinese ideographs drawn by
     湖 远星 to the basic Unifont font files for coverage of
     the General Standard Chinese Characters set.  Also added
     biang glyphs drawn by Ming Fan.
   * Fullwidth CJK characters that appeared to be boldface in
     the range U+FF00..U+FFEF were updated using glyphs from
     Wen Quan Yi, with some modifications.

2022-09-13: Release 15.0.01
   * Update for Unicode 15.0.0 release.
   * "make install" now installs Unifont OpenType font files in the
     same directory as TrueType fonts by default.  The intention is
     to remove TrueType fonts file from the "install" target for the
     Unicode 16.0.0 release unless there is a reason to keep them.
   * Existing Unicode scripts in Unifont maintain complete coverage.
   * Under ConScript Unicode Registry (UCSUR) scripts added:
     - U+E1D0..U+E1FF Syai
     - U+E360..U+E37F Sarkai
     - U+ED80..U+EDAF Iranic.
   * New Unicode 15.0.0 Plane 1 scripts:
     - U+10EFD..U+10EFF Arabic Extended-C
     - U+11B00..U+11B5F Devanagari Extended-A
     - U+11F00..U+11F5F Kawi
     - U+1D2C0..U+1D2DF Kaktovik Numerals
     - U+1E030..U+1E08F Cyrillic Exntended-D
     - U+1E4D0..U+1E4FF Nag Mundari.

2022-06-04: Release 14.0.04

src/johab2syllables.c
/**
   @file johab2syllables.c

   @brief Create the Unicode Hangul Syllables block from component letters.

   This program reads in a "hangul-base.hex" file containing Hangul
   letters in Johab 6/3/1 format and outputs a Unifont .hex format
   file covering the Unicode Hangul Syllables range of U+AC00..U+D7A3.

   @author Paul Hardy

   @copyright Copyright © 2023 Paul Hardy
*/
/*
   LICENSE:

      This program is free software: you can redistribute it and/or modify
      it under the terms of the GNU General Public License as published by
      the Free Software Foundation, either version 2 of the License, or
      (at your option) any later version.

      This program is distributed in the hope that it will be useful,
      but WITHOUT ANY WARRANTY; without even the implied warranty of
      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
      GNU General Public License for more details.

      You should have received a copy of the GNU General Public License
      along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "hangul.h"


/**
   @brief The main function.
*/
int
main (int argc, char *argv[]) {
   int      i;          /* Loop variables     */
   int      arg_count;  /* index into *argv[] */
   unsigned codept;
   unsigned max_codept;
   unsigned char  hangul_base[MAX_GLYPHS][32];
   int      initial, medial, final;         /* Base glyphs for a syllable. */
   unsigned char  syllable[32];        /* Syllable glyph built for output. */

   FILE *infp  = stdin;   /* Input Hangul Johab 6/3/1 file */
   FILE *outfp = stdout;  /* Output Hangul Syllables file  */

   /* Print a help message */
   void print_help ();

   /* Read the file containing Hangul base glyphs. */
   unsigned hangul_read_base8 (FILE *infp, unsigned char hangul_base[][32]);

   /* Given a Hangul Syllables code point, determine component glyphs. */
   void hangul_decompose (unsigned codept, int *, int *, int *);

   /* Given letters in a Hangul syllable, return a glyph. */
   void hangul_syllable (int choseong, int jungseong, int jongseong,
                         unsigned char hangul_base[][32],
                         unsigned char *syllable);


   /*
      If there are command line arguments, parse them.
   */
   arg_count = 1;

   while (arg_count < argc) {
      /* If input file is specified, open it for read access. */
      if (strncmp (argv [arg_count], "-i", 2) == 0) {
         arg_count++;
         if (arg_count < argc) {
            infp = fopen (argv [arg_count], "r");
            if (infp == NULL) {
               fprintf (stderr, "\n*** ERROR: Cannot open %s for input.\n\n",
                        argv [arg_count]);
               exit (EXIT_FAILURE);
            }
         }
      }
      /* If output file is specified, open it for write access. */
      else if (strncmp (argv [arg_count], "-o", 2) == 0) {
         arg_count++;
         if (arg_count < argc) {
            outfp = fopen (argv [arg_count], "w");
            if (outfp == NULL) {
               fprintf (stderr, "\n*** ERROR: Cannot open %s for output.\n\n",
                        argv [arg_count]);
               exit (EXIT_FAILURE);
            }
         }
      }
      /* If help is requested, print help message and exit. */
      else if (strncmp (argv [arg_count], "-h",     2) == 0 ||
               strncmp (argv [arg_count], "--help", 6) == 0) {
         print_help ();
         exit (EXIT_SUCCESS);
      }

      arg_count++;
   }


   /*
      Initialize entire glyph array to zeroes in case the input
      file skips over some code points.
   */
   for (codept = 0; codept < MAX_GLYPHS; codept++) {
      for (i = 0; i < 32; i++) hangul_base[codept][i] = 0;
   }

   /*
      Read the entire "hangul-base.hex" file into an array
      organized as hangul_base [code_point][glyph_byte].
      The Hangul glyphs are 16 columns wide, which is
      two bytes, by 16 rows, for a total of 2 * 16 = 32 bytes.
   */
   max_codept = hangul_read_base8 (infp, hangul_base);
   if (max_codept > 0x8FF) {
      fprintf (stderr, "\nWARNING: Hangul glyph range exceeds PUA space.\n\n");
   }

   /*
      For each glyph in the Unicode Hangul Syllables block,
      form a composite glyph of choseong + jungseong +
      optional jongseong and output it in Unifont .hex format.
   */
   for (codept = 0xAC00; codept < 0xAC00 + 19 * 21 * 28; codept++) {
      hangul_decompose (codept, &initial, &medial, &final);

      hangul_syllable (initial, medial, final, hangul_base, syllable);

      fprintf (outfp, "%04X:", codept);

      for (i = 0; i < 32; i++) {
         fprintf (outfp, "%02X", syllable[i]);
      }
      fputc ('\n', outfp);
   }

   exit (EXIT_SUCCESS);
}


/**
   @brief Print a help message.
*/
void
print_help () {

   printf ("\ngen-hangul [options]\n\n");
   printf ("     Generates Hangul syllables from an input Unifont .hex file encoded\n");
   printf ("     in Johab 6/3/1 format.  The output is the Unicode Hangul Syllables\n");
   printf ("     range, U+AC00..U+D7A3.\n\n");
   printf ("     This program demonstrates forming Hangul syllables without shifting\n");
   printf ("     the final consonant (jongseong) when combined with a vowel having\n");
   printf ("     a long double vertical stroke.  For a program that demonstrtes\n");
   printf ("     shifting jongseong in those cases, see unigen-hangul, which is what\n");
   printf ("     creates the Unifont Hangul Syllables block.\n\n");

   printf ("     This program may be invoked with the following command line options:\n\n");

   printf ("     Option    Parameters    Function\n");
   printf ("     ------    ----------    --------\n");
   printf ("     -h, --help              Print this message and exit.\n\n");
   printf ("     -i        input_file    Unifont hangul-base.hex formatted input file.\n\n");
   printf ("     -o        output_file   Unifont .hex format output file.\n\n");
   printf ("      Example:\n\n");
   printf ("         johab2syllables -i hangul-base.hex -o hangul-syllables.hex\n\n");

   return;
}

